SQL: Update Contract Agreement fails

# Bug Report

## Describe the Bug

When cancelling a *Contract Agreement* on a EDC with SQL extensions, the *Contract Agreement* update fails with an exception (see below for the log output).

### Expected Behavior

Should not fail updating the database entry.

### Observed Behavior

Failed update, see log output below.

## Steps to Reproduce

Steps to reproduce the behavior:
1. Start EDCs with SQL extensions 
2. Setup Contract Agreement. 
3. Cancel Contract Agreement.
4. Look into the log.

## Context Information

- Used version: b71a14c44f3718ce4a2c9b40769bc70d863716ad

## Detailed Description

Log output from Postgres:

```
2022-07-25 08:48:40.949 UTC [86] ERROR:  column "policy" is of type json but expression is of type character varying at character 144
2022-07-25 08:48:40.949 UTC [86] HINT:  You will need to rewrite or cast the expression.
2022-07-25 08:48:40.949 UTC [86] STATEMENT:  UPDATE edc_contract_agreement SET provider_agent_id=$1, consumer_agent_id=$2, signing_date=$3, start_date=$4, end_date=$5, asset_id=$6, policy=$7 WHERE agr_id =$8
```

And the corresponding EDC log:
```
SEVERE 2022-07-25T08:48:45.908075798 Error when processing a command
org.eclipse.dataspaceconnector.spi.persistence.EdcPersistenceException: ERROR: column "policy" is of type json but expression is of type character varying
  Hint: You will need to rewrite or cast the expression.
  Position: 144
    at org.eclipse.dataspaceconnector.sql.SqlQueryExecutor.executeQuery(SqlQueryExecutor.java:52)
    at org.eclipse.dataspaceconnector.sql.contractnegotiation.store.SqlContractNegotiationStore.lambda$upsertAgreement$10(SqlContractNegotiationStore.java:278)
    at org.eclipse.dataspaceconnector.transaction.local.LocalTransactionContext.lambda$execute$0(LocalTransactionContext.java:46)
    at org.eclipse.dataspaceconnector.transaction.local.LocalTransactionContext.execute(LocalTransactionContext.java:63)
    at org.eclipse.dataspaceconnector.transaction.local.LocalTransactionContext.execute(LocalTransactionContext.java:45)
    at org.eclipse.dataspaceconnector.sql.contractnegotiation.store.SqlContractNegotiationStore.upsertAgreement(SqlContractNegotiationStore.java:259)
    at org.eclipse.dataspaceconnector.sql.contractnegotiation.store.SqlContractNegotiationStore.update(SqlContractNegotiationStore.java:217)
    at org.eclipse.dataspaceconnector.sql.contractnegotiation.store.SqlContractNegotiationStore.lambda$save$3(SqlContractNegotiationStore.java:114)
    at org.eclipse.dataspaceconnector.transaction.local.LocalTransactionContext.lambda$execute$0(LocalTransactionContext.java:46)
    at org.eclipse.dataspaceconnector.transaction.local.LocalTransactionContext.execute(LocalTransactionContext.java:63)
    at org.eclipse.dataspaceconnector.transaction.local.LocalTransactionContext.execute(LocalTransactionContext.java:45)
    at org.eclipse.dataspaceconnector.sql.contractnegotiation.store.SqlContractNegotiationStore.save(SqlContractNegotiationStore.java:107)
    at org.eclipse.dataspaceconnector.contract.negotiation.command.handlers.SingleContractNegotiationCommandHandler.handle(SingleContractNegotiationCommandHandler.java:54)
    at org.eclipse.dataspaceconnector.contract.negotiation.command.handlers.SingleContractNegotiationCommandHandler.handle(SingleContractNegotiationCommandHandler.java:31)
    at org.eclipse.dataspaceconnector.spi.command.CommandRunner.runCommand(CommandRunner.java:48)
    at org.eclipse.dataspaceconnector.spi.command.CommandProcessor.processCommandQueue(CommandProcessor.java:46)
    at org.eclipse.dataspaceconnector.contract.negotiation.ProviderContractNegotiationManagerImpl.processCommand(ProviderContractNegotiationManagerImpl.java:261)
    at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
    at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
    at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
    at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
    at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
    at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:258)
    at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:248)
    at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
    at java.base/java.util.stream.ReferencePipeline.count(ReferencePipeline.java:709)
    at org.eclipse.dataspaceconnector.common.statemachine.StateProcessorImpl.process(StateProcessorImpl.java:49)
    at java.base/java.util.stream.ReferencePipeline$5$1.accept(ReferencePipeline.java:231)
    at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
    at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
    at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
    at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
    at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
    at java.base/java.util.stream.LongPipeline.reduce(LongPipeline.java:498)
    at java.base/java.util.stream.LongPipeline.sum(LongPipeline.java:456)
    at org.eclipse.dataspaceconnector.common.statemachine.StateMachineManager.performLogic(StateMachineManager.java:118)
    at org.eclipse.dataspaceconnector.common.statemachine.StateMachineManager.lambda$loop$2(StateMachineManager.java:106)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.postgresql.util.PSQLException: ERROR: column "policy" is of type json but expression is of type character varying
  Hint: You will need to rewrite or cast the expression.
  Position: 144
    at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2468)
    at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2211)
    at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:309)
    at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:446)
    at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:370)
    at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:149)
    at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:138)
    at org.eclipse.dataspaceconnector.sql.SqlQueryExecutor.executeQuery(SqlQueryExecutor.java:50)
    ... 42 more
SEVERE 2022-07-25T08:48:45.909417423 Could not process command [class org.eclipse.dataspaceconnector.contract.negotiation.command.commands.CancelNegotiationCommand], it has exceeded its retry limit, will discard now. Error: [ERROR: column "policy" is of type json but expression is of type character varying   Hint: You will need to rewrite or cast the expression.   Position: 144]
```

## Possible Implementation
_You already know the root cause of the erroneous state and how to fix it? Feel free to share your thoughts._

Fix sql query in extensions/sql/contract-negotiation-store-sql/src/main/java/org/eclipse/dataspaceconnector/sql/contractnegotiation/store/schema/BaseSqlDialectStatements.java

```java
    @Override
     public String getUpdateAgreementTemplate() {
         return format("UPDATE %s SET %s=?, %s=?, %s=?, %s=?, %s=?, %s=?, %s=?%s WHERE %s =?",
                 getContractAgreementTable(), getProviderAgentColumn(), getConsumerAgentColumn(), getSigningDateColumn(),
                 getStartDateColumn(), getEndDateColumn(), getAssetIdColumn(), getPolicyColumn(), getFormatJsonOperator(), getContractAgreementIdColumn());
     }
```

--- a/extensions/sql/contract-negotiation-store-sql/src/main/java/org/eclipse/dataspaceconnector/sql/contractnegotiation/store/schema/BaseSqlDialectStatements.java
+++ b/extensions/sql/contract-negotiation-store-sql/src/main/java/org/eclipse/dataspaceconnector/sql/contractnegotiation/store/schema/BaseSqlDialectStatements.java
@@ -92,9 +92,9 @@ public class BaseSqlDialectStatements implements ContractNegotiationStatements {
 
     @Override
     public String getUpdateAgreementTemplate() {
-        return format("UPDATE %s SET %s=?, %s=?, %s=?, %s=?, %s=?, %s=?, %s=? WHERE %s =?",
+        return format("UPDATE %s SET %s=?, %s=?, %s=?, %s=?, %s=?, %s=?, %s=?%s WHERE %s =?",
                 getContractAgreementTable(), getProviderAgentColumn(), getConsumerAgentColumn(), getSigningDateColumn(),
-                getStartDateColumn(), getEndDateColumn(), getAssetIdColumn(), getPolicyColumn(), getContractAgreementIdColumn());
+                getStartDateColumn(), getEndDateColumn(), getAssetIdColumn(), getPolicyColumn(), getFormatJsonOperator(), getContractAgreementIdColumn());
     }
 
     @Override

--- a/extensions/sql/contract-negotiation-store-sql/src/test/java/org/eclipse/dataspaceconnector/sql/contractnegotiation/store/PostgresContractNegotiationStoreTest.java
+++ b/extensions/sql/contract-negotiation-store-sql/src/test/java/org/eclipse/dataspaceconnector/sql/contractnegotiation/store/PostgresContractNegotiationStoreTest.java
@@ -275,6 +275,27 @@ class PostgresContractNegotiationStoreTest {
         assertThat(store.queryAgreements(query)).isEmpty();
     }
 
+    @Test
+    void create_and_cancel_contractAgreement() {
+        var negotiationId = "test-cn1";
+        var negotiation = createNegotiation(negotiationId);
+        store.save(negotiation);
+
+        // now add the agreement
+        var agreement = createContract("test-ca1");
+        var updatedNegotiation = createNegotiation(negotiationId, agreement);
+
+        store.save(updatedNegotiation);
+        assertThat(store.queryAgreements(QuerySpec.none()))
+                .hasSize(1)
+                .usingRecursiveFieldByFieldElementComparator()
+                .containsExactly(agreement);
+
+        // cancel the agreement
+        updatedNegotiation.transitionError("Cancelled");
+        store.save(updatedNegotiation);
+    }
+
     protected Connection getConnection() {
         try {
             return dataSourceRegistry.resolve(DATASOURCE_NAME).getConnection();

--- a/extensions/sql/contract-negotiation-store-sql/src/test/java/org/eclipse/dataspaceconnector/sql/contractnegotiation/store/SqlContractNegotiationStoreTest.java
+++ b/extensions/sql/contract-negotiation-store-sql/src/test/java/org/eclipse/dataspaceconnector/sql/contractnegotiation/store/SqlContractNegotiationStoreTest.java
@@ -306,6 +306,27 @@ class SqlContractNegotiationStoreTest {
     }
 
     @Test
+    void create_and_cancel_contractAgreement() {
+        var negotiationId = "test-cn1";
+        var negotiation = createNegotiation(negotiationId);
+        store.save(negotiation);
+
+        // now add the agreement
+        var agreement = createContract("test-ca1");
+        var updatedNegotiation = createNegotiation(negotiationId, agreement);
+
+        store.save(updatedNegotiation);
+        assertThat(store.queryAgreements(QuerySpec.none()))
+                .hasSize(1)
+                .usingRecursiveFieldByFieldElementComparator()
+                .containsExactly(agreement);
+
+        // cancel the agreement
+        updatedNegotiation.transitionError("Cancelled");
+        store.save(updatedNegotiation);
+    }
+
+    @Test
     @DisplayName("Should update the agreement when a negotiation is updated")
     void update_whenAgreementExists_shouldUpdate() {
         var negotiationId = "test-cn1";

